extensions/scarb-snforge-test-collector/src/main.rs

```diff
@@ -0,0 +1,88 @@
+use anyhow::{Context, Result};
+use clap::Parser;
+use fs4::FileExt;
+use std::fs::{create_dir_all, File};
+use std::io::Write;
+
+use scarb_metadata::MetadataCommand;
+use scarb_ui::args::PackagesFilter;
+
+use crate::compilation::compile_tests;
+use crate::crate_collection::collect_test_compilation_targets;
+use crate::metadata::{
+    compilation_unit_for_package, corelib_for_compilation_unit, crates_config_for_compilation_unit,
+    dependencies_for_compilation_unit,
+};
+
+mod compilation;
+mod crate_collection;
+mod metadata;
+
+/// Starknet Foundry private extension for compiling test artifacts.
+/// Users should not call it directly.
+#[derive(Parser)]
+#[command(version)]
+struct Args {
+    #[command(flatten)]
+    packages_filter: PackagesFilter,
+}
+
+fn main() -> Result<()> {
+    let args = Args::parse();
+
+    let metadata = MetadataCommand::new().inherit_stderr().exec()?;
+    let selected_packages_metadata = args.packages_filter.match_many(&metadata)?;
+
+    let target_dir = metadata
+        .target_dir
+        .clone()
+        .unwrap_or_else(|| metadata.workspace.root.join("target"));
+    let snforge_target_dir = target_dir.join(&metadata.current_profile).join("snforge");
+    create_dir_all(&snforge_target_dir)?;
+
+    for package_metadata in selected_packages_metadata {
+        let compilation_unit_metadata = compilation_unit_for_package(&metadata, &package_metadata)?;
+
+        let compilation_targets = collect_test_compilation_targets(
+            &package_metadata.name,
+            &package_metadata.root,
+            compilation_unit_metadata.target.source_root(),
+            &compilation_unit_metadata.target.source_path,
+        )?;
+
+        let dependencies = dependencies_for_compilation_unit(compilation_unit_metadata);
+        let corelib_path = corelib_for_compilation_unit(compilation_unit_metadata)?;
+        let crates_config =
+            crates_config_for_compilation_unit(&metadata, compilation_unit_metadata);
```

Comment:
I still don't think this destructuring of `CompilationUnit` is beneficial in any way. I think you would get more flexible api if for instance you wrap `CompilationUnit` into some impl and extract raw paths where needed only, instead of having this series of function calls with long signatures. 

E.g. 
```
#[derive(Clone)]
pub struct CompilationUnit<'a>(&'a CompilationUnitMetadata);

impl CompilationUnit<'_> {
    pub fn dependencies(
        &self,
    ) -> Vec<LinkedLibrary> {
        let dependencies = self.0
            .components
            .iter()
            .filter(|du| &du.name != "core")
            .map(|cu| LinkedLibrary {
                name: cu.name.clone(),
                path: cu.source_root().to_owned().into_std_path_buf(),
            })
            .collect();

        dependencies
    }

    pub fn corelib_path(&self) -> Result<Utf8PathBuf> {
        let corelib = self.0
            .components
            .iter()
            .find(|du| du.name == "core")
            .context("corelib could not be found")?;
        Ok(Utf8PathBuf::from(corelib.source_root()))
    }
}
```

```
pub fn collect_tests(
    crate_name: &str,
    crate_root: &Path,
    lib_content: &str,
    crates_config: AllCratesConfig,
   compilation_unit: CompilationUnit
) -> Result<(Program, Vec<TestCaseRaw>)> {
    let crate_roots: OrderedHashMap<SmolStr, PathBuf> = compilation_unit.dependencies()
        .iter()
        .cloned()
        .map(|source_root| (source_root.name.into(), source_root.path))
        .collect();

    let project_config = ProjectConfig {
        base_path: crate_root.into(),
        corelib: Some(Directory::Real(compilation_unit.corelib_path()?.into())),
```